File: day-3.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <vector>

int main() {
  std::vector<std::vector<bool>> map;

  std::string line;
  while(std::getline(std::cin, line)) {
    auto& row = map.emplace_back();
    row.reserve(line.size());
    for (char c : line) {
      row.push_back(c == '#');
    }
  }

  int slopes[][2] = {{1, 1}, {1, 3}, {1, 5}, {1, 7}, {2, 1}};

  int product = 1;
  for (auto& slope : slopes) {
    int x = 0;
    int trees_encountered = 0;
    for (int i = 0; i < map.size(); i += slope[0]) {
      if (map[i][x]) {
        trees_encountered++;
      }
      x = (x + slope[1]) % map[i].size();
    }
    std::cout << "Slope has " << trees_encountered << " trees" << std::endl;
    product *= trees_encountered;
  }

  std::cout << "Encountered trees: " << product << std::endl;
  return 0;
}

File: 2020/day-3.rs
use std::io::{self, BufRead};

fn main() -> io::Result<()> {
    let mut map: Vec<Vec<bool>> = Vec::new();

    for line in io::stdin().lock().lines() {
        // 1-8 n: dpwpmhknmnlglhjtrbpx
        let line = line.unwrap();
        let row = line.chars().map(|c| c == '#');
        map.push(row.collect());
    }

    let slopes = vec![(1, 1), (1, 3), (1, 5), (1, 7), (2, 1)];
    let mut product = 1;
    for slope in slopes {
        let mut tree_count = 0;
        let mut offset = 0;
        for row in (0..map.len()).step_by(slope.0) {
            if map[row][offset] {
                tree_count += 1;
            }
            offset = (offset + slope.1) % map[row].len();
        }
        println!("Slope {:?} has {} trees", slope, tree_count);
        product *= tree_count;
    }
    println!("Product is {}", product);

    Ok(())
}
